_blog/2024-02-14-writing-async-rust-traits.mdx

```diff
@@ -0,0 +1,318 @@
+
+### Returning “impl Future” in a trait
+
+What about if we don’t know what the future looks like or want to allow the user to decide on the final function signature? For example, if they want to return a box-pinned future? Or just a regular future? We can’t use `async fn` here. In this case, we use `impl Future` as the return type:
+
+```rust
+trait HttpService {
+    async fn fetch(&self, url: Url) -> Body;
+}
+
+// The above would turn into...
+
+trait HttpService {
+    fn fetch(&self, url: Url) -> impl Future<Output = Body>;
+}
+```
+
+A recent change in Rust 1.75 allowed Return Position Impl Trait In Trait (RPITIT - RFCC 1522, 3425).  This is great for us, as it means that we can now use Generic Associated Types (GATs) in a trait! Let’s have a look at the following trait which is just like the previous one, but now has a generic in it:
+
+```rust
+trait HttpService<T> {
+    async fn fetch(&self, url: Url) -> Body<T>;
+}
+
+// The above would turn into...
+
+trait HttpService<T> {
+    type: $<'me, T>: Future<Output = Body<T>>
+    where Self: 'me;
+
+   fn fetch<T>(&self, obj: Url) -> Self::$<'_, T>
+}
+```
+
+Here we have some generic type that implements a generic trait bound, where `Self` is required to be the same lifetime as the inner type.
+
+What about implementing traits for a type, with GATs? You can implement a trait where the type uses them like so:
+
+```rust
+impl<T> Example for HttpService<T> {
+    async fn fetch(&self, obj: Url) -> Body<T> {
+        self.get(obj).await
+    }
+}
+
+// The above would turn into ..
+
+impl<T> Example for HttpService<T> {
+    type $<'me, T> = impl Future<Output = Body<T>> + 'me
+    where Self: 'me;
+
+    fn fetch<T>(&self, obj: Url) -> Self::$<'_, T> {
+        async move { .. }
+    }
+}
+```
+
+You can find more about Generic Associated Types [here.](https://rust-lang.github.io/rfcs/1598-generic_associated_types.html)
+
+### Mixing syntax
+
+Generally speaking, while the above is all well and good you can also mix the syntax to create whatever you need. You don’t necessarily need to 100% implement it one way:
+
+```rust
+trait HttpService: Send {
+    fn fetch(&self, url: Url) -> impl Future<Output = HtmlBody> + Send;
+}
+
+impl HttpService for MyService {
+    async fn fetch(&self, url: Url) -> HtmlBody {
+        self.client.do_fetch(url).await.into_body()
+    }
+}
+```
+
+As you can see here, we add some trait bounds to ensure that the type being implemented always implements `Send` - but then we use async functions inside. 
+
+## Limitations
+
+As you’ve seen, there are a lot of cool different ways you can write async Rust traits! You can do it either the low-level or high-level way, or mix them.
```

Comment:
I feel like we have not made it clear that you can in many cases write async fn in traits natively as of 1.75, see previous comments. Writing your own futures is cool, but it's in most cases a very minor performance optimization, and something the majority of users would not need to do. And following async fn in trait, it should be even more rarely needed. I think we should make this clearer as well.